src/EFCore.PG/Query/ExpressionVisitors/NpgsqlSqlTranslatingExpressionVisitor.cs

```diff
@@ -66,7 +69,7 @@ public class NpgsqlSqlTranslatingExpressionVisitor : SqlTranslatingExpressionVis
 
         /// <inheritdoc />
         protected override Expression VisitSubQuery(SubQueryExpression expression)
-            => base.VisitSubQuery(expression) ?? VisitLikeAnyAll(expression) ?? VisitEqualsAny(expression);
+            => VisitLikeAnyAll(expression) ?? VisitEqualsAny(expression) ?? base.VisitSubQuery(expression);
```

Comment:
Can you give some background on this change? I haven't had much time to review the upstream changes, but IIRC 2.x required this order so we didn't over-glob.

Reply:
Sure. Since 3.0.0 is going to remove client evaluation altogether, EF Core preview4 switches to having it disabled by default. That is, instead of warning when that happens, an exception is thrown - via the same mechanism that already existed in 2.x.

With that setting, `base.VisitSubQuery()` throws an exception - instead of returning null - and doesn't give us an opportunity to try to translate the expression ourselves. This is probably an actual issue in 2.x - we've never tested what happens when client evaluation is disabled - or it could be a problem introduced in the 3.x branch (although I doubt it).

Note that this entire class is going to change significantly once the new query pipeline is merged on the EF Core side - that's when we're going to have some work :) This is why I'm not spending too much time right now about getting things 100% nice/correct.

Reply:
So in general, our logic will have to be to try to translate first ourselves, and only afterwards call the base logic. This also means we have to be extra careful that our own translations are correct and restricted to the cases we want to handle.

Reply:
Makes sense, thanks for the info!